pkg/loadbalancer/experimental/reconciler.go

```diff
@@ -0,0 +1,244 @@
+func (ops *mockOps) Update(ctx context.Context, txn statedb.ReadTxn, fe *Frontend) error {
+	// Gather backends for the service
+	iter := ops.backends.List(txn, BackendServiceIndex.Query(fe.ServiceName))
+	backendRevisions := map[*Backend]statedb.Revision{}
+	for be, rev, ok := iter.Next(); ok; be, rev, ok = iter.Next() {
+		backendRevisions[be] = rev
+	}
+	orderedBackends := sortedBackends(maps.Keys(backendRevisions))
+
+	// Clean up any orphan backends to make room for new backends
+	backendAddrs := sets.New[loadbalancer.L3n4Addr]()
+	for _, be := range orderedBackends {
+		backendAddrs.Insert(be.L3n4Addr)
+	}
+	for _, orphan := range ops.backendsState.updateReferences(fe.Address, backendAddrs) {
+		ops.log.Info("Delete orphan backend", "backend", orphan)
+		ops.backendsState.releaseBackend(orphan)
+	}
+
+	// Update backends that are new or changed.
+	for _, be := range orderedBackends {
+		rev := backendRevisions[be]
+		if !ops.backendsState.needsUpdate(be.L3n4Addr, rev) {
+			continue
+		}
+		ops.log.Info("Update backend", "backend", be)
+		ops.backendsState.updateBackendRevision(be.L3n4Addr, rev)
+	}
+
+	// Assign an identifier for the service. May fail if we have run out of IDs.
+	if fe.ID == 0 {
+		fe.ID = loadbalancer.ID(ops.nextServiceID)
+		ops.nextServiceID++
+	}
+
+	// Update the maglev entries: insert backend IDs into the inner
+	// map and then update the outer map.
+	// TODO
+
+	// Update RevNat
+	ops.log.Info("Update RevNat", "id", fe.ID, "address", fe.Address)
+
+	// Update the master service and backend slots
+	for i, be := range orderedBackends {
+		if be.State != loadbalancer.BackendStateActive {
+			break
+		}
+		ops.log.Info("Update service", "id", fe.ID, "slot", i+1)
+	}
+	ops.log.Info("Update master service", "id", fe.ID)
+	ops.log.Info("Cleanup service slots", "id", fe.ID, "active", numActive(orderedBackends), "previous", ops.numBackends[fe.Address])
+	ops.numBackends[fe.Address] = numActive(orderedBackends)
+
+	return nil
+}
+
+var _ reconciler.Operations[*Frontend] = &mockOps{}
+
+// backendsState tracks what backends exist and how they're referenced by services.
+type backendsState struct {
+	revisions map[loadbalancer.L3n4Addr]statedb.Revision
+
+	// references maps service frontend to the associated backends
+	references map[loadbalancer.L3n4Addr]sets.Set[loadbalancer.L3n4Addr]
+	refCounts  map[loadbalancer.L3n4Addr]int
+}
+
+func (s *backendsState) updateReferences(frontend loadbalancer.L3n4Addr, backends sets.Set[loadbalancer.L3n4Addr]) (orphans []loadbalancer.L3n4Addr) {
+	newRefs := backends.Clone()
+	if oldRefs, ok := s.references[frontend]; ok {
+		for addr := range oldRefs {
+			if newRefs.Has(addr) {
+				newRefs.Delete(addr)
+				continue
+			}
+
+			count := s.refCounts[addr] - 1
+			if count <= 0 {
+				//log.Infof("Backend %s is now an orphan", addr.StringWithProtocol())
+				orphans = append(orphans, addr)
+			} else {
+				//log.Infof("Backend %s ref count now %d", addr.StringWithProtocol(), count)
```

Comment:
nit: We'd usually drop inactive code lines like this.
```suggestion
				orphans = append(orphans, addr)
			} else {
```